x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,188 @@ import (
-			utils.Attribute{Key: "cu_to_be_added", Value: strconv.FormatUint(cu, 10)})
+			utils.Attribute{Key: "sub_block", Value: strconv.FormatUint(entryBlock, 10)},
+			utils.Attribute{Key: "current_cu", Value: strconv.FormatUint(cu, 10)},
+			utils.Attribute{Key: "cu_to_be_added", Value: strconv.FormatUint(cuToAdd, 10)})
 	}
 	return nil
 }
+
+// GetAllSubTrackedCuIndices gets all the trackedCu entries that are related to a specific subscription
+func (k Keeper) GetAllSubTrackedCuIndices(ctx sdk.Context, sub string) []string {
+	return k.cuTrackerFS.GetAllEntryIndicesWithPrefix(ctx, sub)
+}
+
+// removeCuTracker removes a trackedCu entry
+func (k Keeper) resetCuTracker(ctx sdk.Context, sub string, info trackedCuInfo, shouldRemove bool) error {
+	key := types.CuTrackerKey(sub, info.provider, info.chainID)
+	if shouldRemove {
+		return k.cuTrackerFS.DelEntry(ctx, key, uint64(ctx.BlockHeight()))
+	}
+
+	var trackedCu types.TrackedCu
+	found := k.cuTrackerFS.FindEntry(ctx, key, info.block, &trackedCu)
+	if !found {
+		return utils.LavaFormatError("cannot reset CU tracker", fmt.Errorf("trackedCU entry is not found"),
+			utils.Attribute{Key: "sub", Value: sub},
+			utils.Attribute{Key: "provider", Value: info.provider},
+			utils.Attribute{Key: "chainID", Value: info.chainID},
+			utils.Attribute{Key: "block", Value: strconv.FormatUint(info.block, 10)},
+		)
+	}
+
+	trackedCu.Cu = 0
+	k.cuTrackerFS.ModifyEntry(ctx, key, info.block, &trackedCu)
+	return nil
+}
+
+type trackedCuInfo struct {
+	provider  string
+	chainID   string
+	trackedCu uint64
+	block     uint64
+}
+
+func (k Keeper) getSubTrackedCuInfo(ctx sdk.Context, sub string) (trackedCuList []trackedCuInfo, totalCuUsedBySub uint64) {
+	keys := k.GetAllSubTrackedCuIndices(ctx, sub)
+
+	for _, key := range keys {
+		_, provider, chainID := types.DecodeCuTrackerKey(key)
+		cu, entryBlock, found, _ := k.GetTrackedCu(ctx, sub, provider, chainID)
+		if !found {
+			utils.LavaFormatWarning("cannot remove cu tracker", legacyerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "block", Value: strconv.FormatUint(entryBlock, 10)},
+			)
+		}
+		trackedCuList = append(trackedCuList, trackedCuInfo{
+			provider:  provider,
+			trackedCu: cu,
+			chainID:   chainID,
+			block:     entryBlock,
+		})
+		totalCuUsedBySub += cu
+	}
+
+	return trackedCuList, totalCuUsedBySub
+}
+
+// remove only before the sub is deleted
+func (k Keeper) RewardAndResetCuTracker(ctx sdk.Context, cuTrackerTimerKeyBytes []byte) {
+	sub, shouldRemove := types.DecodeCuTrackerTimerKey(string(cuTrackerTimerKeyBytes))
+	trackedCuList, totalCuUsedBySub := k.getSubTrackedCuInfo(ctx, sub)
+
+	for _, trackedCuInfo := range trackedCuList {
+		trackedCu := trackedCuInfo.trackedCu
+		provider := trackedCuInfo.provider
+		chainID := trackedCuInfo.chainID
+		block := trackedCuInfo.block
+
+		plan, err := k.GetPlanFromSubscription(ctx, sub, block)
```

Comment:
this can be done before the for loop since it will always be the same

Reply:
I know, but it's less readable if I put it above, no?

Reply:
done. see https://github.com/lavanet/lava/commit/36c4bd192408759547b2588605992db3d3b23f21